### PR TITLE
[ISSUE #131] Support Snapshot for Dledger

### PR DESCRIPTION
**Pull request by tsunghanjacktsai**
## Test List

- [x] Unit testing, functional testing of each interface.
- [x] Test loading snapshot after node restart.
- [x] Test starting the DLedger under normal mode without a state machine.
- [x] Test if the number of reserved snapshots meets the specified requirements.
- [x] Test if previous snapshots can be loaded once the current one fails.

### CODE REVIEW

File: src/main/java/io/openmessaging/storage/dledger/snapshot/file/FileSnapshotStorage.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                if (!fileName.startsWith(SnapshotFile.SNAPSHOT_DIR_PREFIX)) {
+                    continue;
+                }
+                final long snapshotIndex = Long.parseLong(fileName.substring(SnapshotFile.SNAPSHOT_DIR_PREFIX.length()));
+                snapshotIdx.add(snapshotIndex);
+            }
+            if (!snapshotIdx.isEmpty()) {
+                Collections.sort(snapshotIdx);
+                this.lastSnapshotIndex = snapshotIdx.get(snapshotIdx.size() - 1);
+            }
```
Comment (RongtongJin) [inline src/main/java/io/openmessaging/storage/dledger/snapshot/file/FileSnapshotStorage.java:79]: 可以再上一个循环就把最后一个lastSnapshotIndex找出来，不需要放在Collections里排序

File: src/main/java/io/openmessaging/storage/dledger/store/file/DLedgerMmapFileStore.java [hunk 2]
```diff
@@ -85,15 +85,19 @@ public DLedgerMmapFileStore(DLedgerConfig dLedgerConfig, MemberState memberState
...
 
     @Override
     public void startup() {
         load();
         recover();
         flushDataService.start();
-        cleanSpaceService.start();
+        if (!dLedgerConfig.isStatemachineEnable()) {
+            cleanSpaceService.start();
+        }
```
Comment (RongtongJin) [inline src/main/java/io/openmessaging/storage/dledger/store/file/DLedgerMmapFileStore.java:100]: shutdown 也要判断是否是 null

Comment (RongtongJin) [reply src/main/java/io/openmessaging/storage/dledger/store/file/DLedgerMmapFileStore.java:100]: 应该是判断是否是null而不是isStatemachineEnable

**Review by RongtongJin — CHANGES_REQUESTED** [review]

File: src/main/java/io/openmessaging/storage/dledger/DLedgerServer.java [hunk 3]
```diff
@@ -136,10 +142,13 @@ public MemberState getMemberState() {
...
     public void registerStateMachine(final StateMachine fsm) {
         final StateMachineCaller fsmCaller = new StateMachineCaller(this.dLedgerStore, fsm, this.dLedgerEntryPusher);
-        fsmCaller.start();
+        fsmCaller.registerSnapshotManager(new FileSnapshotManager(fsmCaller, this.dLedgerConfig));
         this.fsmCaller = Optional.of(fsmCaller);
         // Register state machine caller to entry pusher
         this.dLedgerEntryPusher.registerStateMachine(this.fsmCaller);
+        if (this.dLedgerConfig.getStoreType().equals(DLedgerConfig.FILE)) {
+            dLedgerStore.setEnableStateMachine(true);
+        }
```
Comment (RongtongJin) [inline src/main/java/io/openmessaging/storage/dledger/DLedgerServer.java:151]: 需要在代码中保证必须是先注册，再启动，否则抛出错误

File: src/main/java/io/openmessaging/storage/dledger/statemachine/StateMachineCaller.java [hunk 4]
```diff
@@ -166,10 +179,61 @@ private void doCommitted(final long committedIndex) {
...
+        if (res == 0) {
+            return Long.compare(this.lastAppliedIndex.get(), snapshotIndex);
+        } else {
+            return res;
+        }
+    }
+
+    private void doSnapshotSave(SaveSnapshotHook saveSnapshotAfter) {
+        // Build and save snapshot meta
+        SnapshotMeta snapshotMeta = new SnapshotMeta(this.lastAppliedIndex.get(), this.lastAppliedTerm);
```
Comment (RongtongJin) [inline src/main/java/io/openmessaging/storage/dledger/statemachine/StateMachineCaller.java:230]: 如何保证this.lastAppliedIndex.get(), this.lastAppliedTerm和SaveSnapshotHook saveSnapshotAfter中的dledger entry的index和term一致？因为中间可能有其他apply Task进入

**Comment by RongtongJin** [comment]
Hi, @hzh0425  @mxsm  @TheR1sing3un Plz help review this PR if you have time.

**Comment by TheR1sing3un** [comment]
Hi~Please resolve the conflicts if you have time~

File: dledger/src/main/java/io/openmessaging/storage/dledger/statemachine/StateMachineCaller.java [hunk 5]
```diff
@@ -143,6 +156,9 @@ public void run() {
     }
 
     private void doCommitted(final long committedIndex) {
+        if (this.error != null || this.snapshotManager.isLoadingSnapshot()) {
```
[the remark this answers is not in the document]
Comment (tsunghanjacktsai) [reply dledger/src/main/java/io/openmessaging/storage/dledger/statemachine/StateMachineCaller.java]: @RongtongJin Done.

File: dledger/src/main/java/io/openmessaging/storage/dledger/snapshot/SnapshotManager.java [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+            return;
+        }
+        // Start saving snapshot
+        this.savingSnapshot = true;
+        SaveSnapshotAfterHook saveSnapshotAfter = new SaveSnapshotAfterHook(writer, dLedgerEntry);
+        if (!this.dLedgerServer.getFsmCaller().onSnapshotSave(saveSnapshotAfter)) {
+            logger.error("Unable to call statemachine onSnapshotSave");
+            saveSnapshotAfter.doHook(SnapshotStatus.FAIL);
+        }
+    }
```
Comment (mxsm) [inline dledger/src/main/java/io/openmessaging/storage/dledger/snapshot/SnapshotManager.java:146]: Concurrency issue

**Review by RongtongJin — APPROVED** [review]

**Pull request merged**